### PR TITLE
GEOMESA-1723 HBase server-side filter hooks

### PR DESCRIPTION
**Pull request by MutahirKazmi**
This commit adds a client side filtering mechanism to HBase. It enables the system to have the ability to determine the type of filter that needs to be applied from the Hint. 

Tests has been run for all the system using (mvn clean install) including running HBaseDatastoreTest separately.

Signed-off-by: Mutahir <kazmi.mutahir@gmail.com>

### CODE REVIEW

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala [hunk 1]
```diff
@@ -55,10 +56,11 @@ case class EmptyPlan(filter: HBaseFilterStrategyType) extends HBaseQueryPlan {
 case class ScanPlan(filter: HBaseFilterStrategyType,
                     table: TableName,
                     ranges: Seq[Scan],
+                    remoteFilters: Seq[HBaseFilter] = Nil,
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: this needs to be added to HBaseQueryPlan, and then implemented in EmptyPlan and GetPlan

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala [hunk 2]
```diff
@@ -30,15 +31,21 @@ object HBaseFeatureIndex extends HBaseIndexManagerType {
   override val CurrentIndices: Seq[HBaseFeatureIndex] =
     Seq(HBaseZ3Index, HBaseXZ3Index, HBaseZ2Index, HBaseXZ2Index, HBaseIdIndex, HBaseAttributeIndex)
 
-  val DataColumnFamily = Bytes.toBytes("d")
-  val DataColumnFamilyDescriptor = new HColumnDescriptor(DataColumnFamily)
+  val DataColumnFamily = new Text("d")
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: could you revert these back to byte arrays? doesn't seem to be used as texts anywhere

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala [hunk 3]
```diff
@@ -1,24 +1,25 @@
...
-* which accompanies this distribution and is available at
-* http://www.opensource.org/licenses/apache2.0.php.
-*************************************************************************/
+/** *********************************************************************
+  * Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+  * All rights reserved. This program and the accompanying materials
+  * are made available under the terms of the Apache License, Version 2.0
+  * which accompanies this distribution and is available at
+  * http://www.opensource.org/licenses/apache2.0.php.
+  * ************************************************************************/
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: this header formatting change is going to cause license check failure

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala [hunk 4]
```diff
@@ -55,10 +56,11 @@ case class EmptyPlan(filter: HBaseFilterStrategyType) extends HBaseQueryPlan {
...
                     resultsToFeatures: Iterator[Result] => Iterator[SimpleFeature]) extends HBaseQueryPlan {
   override def scan(ds: HBaseDataStore): CloseableIterator[SimpleFeature] = {
-    val results = new BatchScan(ds.connection, table, ranges, ds.config.queryThreads, 100000)
-    SelfClosingIterator(resultsToFeatures(results), results.close)
+      val results = new BatchScan(ds.connection, table, ranges, ds.config.queryThreads, 100000, remoteFilters)
```
Comment (jahhulbert-ccri) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: picky - keep 2 spaces for indent ?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala (hunk 1 above)
Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: Added that to both as well.

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala (hunk 3 above)
Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: Made the change.

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala (hunk 2 above)
Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: Done

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala (hunk 4 above)
Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: Done

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala [hunk 5]
```diff
@@ -48,23 +50,26 @@ object HBaseQueryPlan {
...
-    val results = new HBaseBatchScan(ds.connection, table, ranges, ds.config.queryThreads, 100000)
+    val results = new HBaseBatchScan(ds.connection, table, ranges, ds.config.queryThreads, 100000, remoteFilters)
     SelfClosingIterator(resultsToFeatures(results), results.close)
   }
 }
 
 case class GetPlan(filter: HBaseFilterStrategyType,
                    table: TableName,
                    ranges: Seq[Get],
+                   remoteFilters: Seq[HBaseFilter] = Nil,
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: this needs to be applied to the `get` on line 77 I think?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala [hunk 6]
```diff
@@ -41,10 +44,16 @@ object HBaseFeatureIndex extends HBaseIndexManagerType {
...
+  case class ScanConfig(hbaseFilters: Seq[HBaseFilter],
+                        columnFamily: Array[Byte],
+                        entriesToFeatures: Iterator[Result] => Iterator[SimpleFeature],
+                        reduce: Option[(CloseableIterator[SimpleFeature]) => CloseableIterator[SimpleFeature]])
+
 }
 
 trait HBaseFeatureIndex extends HBaseFeatureIndexType
-    with IndexAdapter[HBaseDataStore, HBaseFeature, Mutation, Query] with ClientSideFiltering[Result] {
+  with IndexAdapter[HBaseDataStore, HBaseFeature, Mutation, Query] with ClientSideFiltering[Result] {
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: [picky] revert indentation

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala [hunk 7]
```diff
@@ -147,4 +156,33 @@ trait HBaseFeatureIndex extends HBaseFeatureIndexType
...
+    * Sets up everything needed to execute the scan - iterators, column families, deserialization, etc
+    *
+    * @param sft    simple feature type
+    * @param filter hbase filter strategy type
+    * @param hints  query hints
+    * @param ecql   secondary filter being applied, if any
+    * @param dedupe scan may have duplicate results or not
+    * @return
+    */
+
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: [picky] remove empty line between scaladocs and method

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala (hunk 6 above)
Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: I made this indentation based on @jahhulbert-ccri . Should I keep the one in Red or Green?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala [hunk 8]
```diff
@@ -41,10 +44,16 @@ object HBaseFeatureIndex extends HBaseIndexManagerType {
 
   val DataColumnQualifier = Bytes.toBytes("d")
   val DataColumnQualifierDescriptor = new HColumnDescriptor(DataColumnQualifier)
+
+  case class ScanConfig(hbaseFilters: Seq[HBaseFilter],
+                        columnFamily: Array[Byte],
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: column family is already incorporated in the `ranges`, so I think we can remove it here as it's not used

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala [hunk 9]
```diff
@@ -41,10 +44,16 @@ object HBaseFeatureIndex extends HBaseIndexManagerType {
...
+                        entriesToFeatures: Iterator[Result] => Iterator[SimpleFeature],
+                        reduce: Option[(CloseableIterator[SimpleFeature]) => CloseableIterator[SimpleFeature]])
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: reduce is also currently incorporated into `entriesToFeatures`, and isn't used, so we should remove it for now

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala (hunk 5 above)
Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: I am not sure what you are implying here. Place the remoteFilters on:

SelfClosingIterator(resultsToFeatures(get.get(ranges).iterator), get.close)

?

Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: yeah, you applied the server-side filters in `HBaseBatchScan`, which works for the `ScanPlan`, but the `GetPlan` never has the server-side filter applied.

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala (hunk 6 above)
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: haha, sorry for jerking you back and forth... I think the original line in red is correct (indent one extra for continuation of a method signature)

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala (hunk 5 above)
Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: I will talk to @jnh5y trying to see what I can do for this.

Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: thanks, once you get that sorted out I think we can merge

Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: Ok, so the variables here have confusing names...  'get' is a table, and it has a 'get' method on it.

Some class needs to take the 'remoteFilters' (if configured) in as input.  It is either the HBase Connection, the Table, the Get, or the Get's call to get.

Does that make sense?  If you don't pass the remoteFilters, there's no filtering:(.

Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: get it? :D

Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: @elahrvivaz i hope so. Just pushed.

Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: looks good - I'll merge

Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: (was joking about the get)

**Comment by MutahirKazmi** [comment]
Is there anything required from my end since @elahrvivaz said it is good to go.

**Comment by elahrvivaz** [comment]
I think it's good. I'll merge today unless someone objects

**Comment by jahhulbert-ccri** [comment]
:shipit:

**Pull request merged**